### PR TITLE
feat(mcp): gittensory_check_before_start (pre-start duplicate/solvability check)

### LINKED ISSUE #545: feat(mcp): gittensory_check_before_start (pre-start duplicate/solvability check)
Parent roadmap: #525
Parent phase: #544

## Background

Today collision/duplicate detection only runs *after* a branch exists (the `preflight_*` tools). Miners routinely waste effort building on an issue that already has a linked PR or isn't solvable — a top source of "slop" from the maintainer's view.

## Goal

New MCP tool `gittensory_check_before_start` that answers, *before any code is written*: is this issue already claimed/solved? is a duplicate cluster forming? is it a valid target?

## Current Behavior

No pre-start check exists; collision runs only post-branch.

## Desired Behavior

Input `{ owner, repo, issueNumber | title | plannedPaths }`. Returns claim/solved status, duplicate-cluster risk, and a go/raise/avoid recommendation with reasons. Reuses `buildCollisionReport` (`engine.ts:756`), `buildIssueQualityReport` (`:2717`), `buildIssueDiscoveryLifecycleReport` (`:2792`).

## Implementation Requirements

- Register in `src/mcp/server.ts`; expose in `packages/gittensory-mcp/bin/gittensory-mcp.js`. Add an `outputSchema`. Metadata only, no writes.

## Public/Private Output Boundaries

- Public-safe reasons only; no reward/score/trust language. Never reveals other contributors' private context.

## Acceptance Criteria

- An already-linked or unsolvable issue returns avoid/raise with specific reasons; a clean target returns go.

## Testing Requirements

- `npm run test:ci`, 97%+ coverage. Unit tests over claimed/solved/duplicate/clean fixtures; forbidden-term assertions.

### PR DESCRIPTION
**Pull request by oktofeesh1**
Fixes #545

## What

Adds `gittensory_check_before_start`, a pre-start MCP tool that answers — **before any code is written** — whether an issue is already claimed or solved, whether a duplicate cluster is forming, and whether it is a valid target. It returns a `go` / `raise` / `avoid` recommendation with public-safe reasons from cached metadata. Metadata only; no GitHub writes.

This moves collision/solvability detection earlier than the existing `preflight_*` tools (which only run after a branch exists), addressing a top source of wasted effort / slop.

## How

- **`src/signals/engine.ts`** — new deterministic `buildPreStartCheck` builder that **reuses the existing reports**: `buildCollisionReport`, `buildIssueQualityReport`, and `buildIssueDiscoveryLifecycleReport`. It resolves the target by `issueNumber`, fuzzy `title` match (token Jaccard), or `plannedPaths` overlap with recently merged work, then derives `claimStatus` (unclaimed/claimed/solved/unknown), `duplicateClusterRisk`, and the recommendation.
- **Public-safe by construction** — every reason/blocker is routed through `sanitizePublicComment` (fail-closed); no reward/score/trust language, and other contributors' private context is never surfaced.
- **`src/mcp/server.ts`** — registers `gittensory_check_before_start` with an `inputSchema` and an `outputSchema`.
- **`src/api/routes.ts`** — `POST /v1/repos/:owner/:repo/check-before-start` backing the local CLI tool (mirrors the `issue-quality` route's auth).
- **`packages/gittensory-mcp/bin/gittensory-mcp.js`** — exposes the tool, proxying to the API route.

## Output boundaries

Public-safe reasons only; metadata only, no writes. The tool never reveals reward/score/trust internals or other contributors' private context.

## Tests

- `test/unit/pre-start-check.test.ts` — go/raise/avoid, claimed (open linked PR), solved (merged work), high duplicate-cluster risk, duplicate/invalid labels, needs-proof, not-found, fuzzy-title resolution, title no-match, direct-PR lane, planned-path overlap, and forbidden-term assertions on all output.
- `test/unit/mcp-output-schemas.test.ts` — output-schema discovery + structured-content call.
- `test/integration/api.test.ts` — route success, invalid body (400), and session-forbidden (403).

`npm run validate` (typecheck + coverage) is green at the 97%+ branch threshold; `build:mcp`, `test:mcp-pack`, `test:workers`, and `ui:openapi:check` also pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ghost** [comment]
Superagent didn't find any vulnerabilities or security issues in this PR.

**Comment by ghost** [comment]
**gittensory** · advisory review

Reviewed **7** changed file(s) — two independent AI reviewers.

**Suggested action:** ✅ **Safe to merge** — both reviewers found no blocking issues.

**What changed**
- Added checkBeforeStartShape schema to CLI and server.
- Implemented new API route `/v1/repos/:owner/:repo/check-before-start`.
- Added MCP tool registration and handler `checkBeforeStart`.
- Implemented `buildPreStartCheck` with public‑safe sanitization.
- Updated output schema list and added related unit/integration tests.
- Added new test suite `pre-start-check.test.ts` covering many scenarios.

**Reviewer A · `gpt-oss-120b`** — recommends ✅ `merge`
This PR introduces a new pre‑start check tool (gittensory_check_before_start) with API route, MCP registration, implementation, and comprehensive tests. The implementation respects the public‑sanitizer and aligns with existing architecture.

_Suggestions_
- Consider adding the max title length constraint to the CLI input schema for consistency with the server validation.
- Add a unit test covering title length validation to ensure the tool rejects overly long titles before hitting the server.

_Worth double-checking_
- CLI schema allows titles longer than the server's max length, which will result in a 400 error at runtime.
- Ensure the imported sanitizePublicComment path is correct and that the function does not throw on empty strings.

**Reviewer B · `nemotron-3-120b-a12b`** — recommends ✅ `merge`
This PR adds a new MCP tool 'gittensory_check_before_start' for pre-start duplicate/solvability checks. The changes span MCP tool registration, API route, business logic in signals/engine, and comprehensive tests. The implementation follows project conventions: uses proper validation, authentication, sanitizes public output, and includes unit/integration tests. The code is clean, well-tested, and addresses a clear feature need.

_Suggestions_
- Consider adding a comment in src/signals/engine.ts explaining why plannedPaths are lowercased (for case-insensitive file system matching).
- In test/unit/pre-start-check.test.ts, the test 'returns go for planned paths with no overlapping merged work' could assert that duplicateClusterRisk is 'none' for clarity.
- The MCP tool description in src/mcp/server.ts could explicitly mention that no GitHub writes occur (already implied by 'No GitHub writes' but could be emphasized).

_Worth double-checking_
- Ensure the sanitizePublicComment function is indeed imported and available in src/signals/engine.ts (it is, via the added import).
- Verify that the new route doesn't inadvertently expose private data - the use of buildPreStartCheck with sanitization appears correct.
- Check that the MCP tool's output schema matches what's actually returned (it does, based on the handler).

**Comment by ghost** [comment]
> [!NOTE]
> ## Gittensory Gate skipped
> PR closed before full evaluation. No late first comment was created.
>
> | Signal | Result | Evidence | Action |
> | --- | --- | --- | --- |
> | Gate result | ⚠️ Skipped | JSONbored/gittensory#621 is no longer open. | No action. |

---
💰 **Earn for open-source contributions like this.** [Gittensor](https://gittensor.io) lets GitHub contributors earn for the work they already do — [register to start earning →](https://gittensor.io/miners/repository?name=JSONbored%2Fgittensory&tab=miners).

Checked by [Gittensory](https://gittensory.aethereal.dev), a quiet PR intelligence layer for OSS maintainers.

**Review by JSONbored — APPROVED** [review]

**Review by JSONbored — CHANGES_REQUESTED** [review]
fix failing CI, please.

**Review by ghost — COMMENTED** [review]
**Gittensory review** · _advisory — a maintainer merges_

**3** actionable · **5** nitpick(s) · 8 file(s) · two independent AI reviewers

> **Suggested action:** ✅ **Safe to merge** — both reviewers found no blocking issues.

<details>
<summary>📋 Walkthrough</summary>

Adds a new `gittensory_check_before_start` tool that evaluates issue claim status, duplicate risk, and other signals before coding begins, exposing a POST `/check-before-start` endpoint and accompanying MCP method, plus comprehensive unit and integration tests.

**Changes**

| File | Summary |
| --- | --- |
| `packages/gittensory-mcp/bin/gittensory-mcp.js` | Defines input shape for the new tool and registers the `gittensory_check_before_start` tool in the CLI. |
| `src/api/routes.ts` | Adds request schema and POST `/v1/repos/:owner/:repo/check-before-start` handler with auth checks. |
| `src/mcp/server.ts` | Introduces input/output schemas, registers the new tool, and implements `checkBeforeStart` method. |
| `src/signals/engine.ts` | Implements `buildPreStartCheck` function and related types/constants for the pre‑start analysis. |
| `test/integration/api.test.ts` | Tests successful, invalid, and forbidden calls to the new endpoint. |
| `test/unit/mcp-output-schemas.test.ts` | Ensures the new tool appears in the output‑schema list and validates its response shape. |
| `test/unit/mcp-upstream.test.ts` | Adds test confirming session actors cannot run checks on inaccessible repos. |
| `test/unit/pre-start-check.test.ts` | Provides extensive unit tests covering various scenarios of the pre‑start check logic. |

</details>

<details>
<summary>🔍 Reviewer notes</summary>

**Reviewer A · `gpt-oss-120b`** — recommends ✅ `merge`
This PR introduces a new pre‑start duplicate/solvability check tool, complete with API route, MCP registration, implementation, and extensive tests. The changes are well‑scoped, respect the public‑private boundaries, and appear technically sound.

_Suggestions_
- Consider adding the same max‑length constraints to the CLI `checkBeforeStartShape` as used in the server for consistency.
- Document that the `report` field in the tool output may contain raw data; ensure it does not include forbidden public terms.

_Worth double-checking_
- The `report` object is returned publicly; verify it cannot contain forbidden terms from issue titles or other metadata.
- Ensure the `sanitizePublicComment` import is present in `engine.ts` (it is used elsewhere but double‑check).

**Reviewer B · `mistral-small-3.1-24b-instruct`** — recommends ✅ `merge`
This PR adds a new pre-start check feature to the Gittensory MCP. It includes a new tool, API route, and corresponding tests. The implementation is thorough and well-documented, with clear separation of concerns and adherence to project conventions.

_Suggestions_
- Consider adding a brief description of the new tool in the README or documentation to inform users about its purpose and usage.
- Ensure that the new API route is documented in the API documentation to make it discoverable for developers.

_Worth double-checking_
- Verify that the new tool does not introduce any performance regressions, especially when handling large repositories or a high number of issues/PRs.
- Double-check that the sanitization of public comments is comprehensive and covers all potential edge cases.
- Ensure that the new tests cover all possible scenarios and edge cases, including invalid inputs and error handling.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

Copy this to your coding agent to fix everything flagged:

```
Address the following in this PR:

1. src/signals/engine.ts (line 2951): The tokenization and matching logic for issue titles could be optimized to handle large datasets more efficiently.
2. src/signals/engine.ts (line 2970): The magic number 0.5 used for title matching should be extracted to a named constant for better readability and maintainability. Suggested fix: const TITLE_MATCH_THRESHOLD = 0.5;
3. src/signals/engine.ts (line 3047): The status strings like 'go', 'raise', 'avoid' should be replaced with enum values for type safety and better code readability. Suggested fix: enum PreStartCheckRecommendation { GO = 'go', RAISE = 'raise', AVOID = 'avoid' }
4. test/unit/pre-start-check.test.ts (line 104): The test names should follow a consistent naming convention for better readability and maintainability. Suggested fix: it('should return go for a clean, unclaimed, actionable issue', () => { ... }

Nitpicks:
- Consider adding the same max‑length constraints to the CLI `checkBeforeStartShape` as used in the server for consistency.
- Document that the `report` field in the tool output may contain raw data; ensure it does not include forbidden public terms.
- Consider adding a brief description of the new tool in the README or documentation to inform users about its purpose and usage.
- Ensure that the new API route is documented in the API documentation to make it discoverable for developers.
```

</details>

---
💰 Earn for open-source contributions like this — [Gittensor](https://gittensor.io) lets contributors earn for the work they already do. · _Advisory review by Gittensory._

File: src/signals/engine.ts [hunk 1]
```diff
@@ -2919,6 +2919,184 @@ export function buildLinkedIssueValidation(
...
+  found: boolean;
+  claimStatus: PreStartCheckClaimStatus;
+  lifecycle?: IssueDiscoveryLifecycleState | undefined;
+  issueQualityStatus?: "ready" | "needs_proof" | "hold" | "do_not_use" | undefined;
+  duplicateClusterRisk: DuplicateClusterRisk;
+  recommendation: PreStartCheckRecommendation;
+  reasons: string[];
+  blockers: string[];
+  summary: string;
+};
```
Comment (ghost) [inline src/signals/engine.ts:2951]: **⚠️ Potential issue** · Potential Performance Issue

The tokenization and matching logic for issue titles could be optimized to handle large datasets more efficiently.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 2951, The tokenization and matching logic for issue titles could be optimized to handle large datasets more efficiently.
```

</details>

File: src/signals/engine.ts [hunk 2]
```diff
@@ -2919,6 +2919,184 @@ export function buildLinkedIssueValidation(
...
+export function buildPreStartCheck(
+  repo: RepositoryRecord | null,
+  issues: IssueRecord[],
+  pullRequests: PullRequestRecord[],
+  recentMergedPullRequests: RecentMergedPullRequestRecord[],
+  fullName: string,
+  target: PreStartCheckTarget,
+): PreStartCheckReport {
+  const lane = buildLaneAdvice(repo, fullName);
+  const collisions = buildCollisionReport(fullName, issues, pullRequests, recentMergedPullRequests);
```
Comment (ghost) [inline src/signals/engine.ts:2975]: **🛠️ Refactor suggestion** · Extract Magic Numbers

The magic number 0.5 used for title matching should be extracted to a named constant for better readability and maintainability.

```suggestion
const TITLE_MATCH_THRESHOLD = 0.5;
```

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 2970, The magic number 0.5 used for title matching should be extracted to a named constant for better readability and maintainability. Apply: const TITLE_MATCH_THRESHOLD = 0.5;
```

</details>

File: src/signals/engine.ts [hunk 3]
```diff
@@ -2919,6 +2919,184 @@ export function buildLinkedIssueValidation(
...
+    );
+  }
+  if (claimStatus === "solved") blockers.push("This issue already has merged or validated solving work; new work would likely duplicate it.");
+  if (claimStatus === "claimed") {
+    blockers.push(
+      resolvedNumber != null
+        ? "Open PR work already references this issue; coordinate or pick a different target to avoid a collision."
+        : "Recently merged work already touched one or more of these paths; confirm this is not a duplicate before starting.",
+    );
+  }
```
Comment (ghost) [inline src/signals/engine.ts:3053]: **🛠️ Refactor suggestion** · Use Enums for Statuses

The status strings like 'go', 'raise', 'avoid' should be replaced with enum values for type safety and better code readability.

```suggestion
enum PreStartCheckRecommendation { GO = 'go', RAISE = 'raise', AVOID = 'avoid' }
```

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 3047, The status strings like 'go', 'raise', 'avoid' should be replaced with enum values for type safety and better code readability. Apply: enum PreStartCheckRecommendation { GO = 'go', RAISE = 'raise', AVOID = 'avoid' }
```

</details>

File: test/unit/pre-start-check.test.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    expect(report.summary.startsWith("Go")).toBe(true);
+    expect(report.reasons.length).toBeGreaterThan(0);
+    assertPublicSafe(report);
+  });
+
+  it("avoids an issue already claimed by an open linked PR", () => {
+    const r = repo("owner/repo");
+    const issues = [issue(1, "Fix parser crash on empty input handling")];
+    const prs = [pr(10, "Fix parser crash", { linkedIssues: [1] })];
+    const report = buildPreStartCheck(r, issues, prs, [], "owner/repo", { issueNumber: 1 });
```
Comment (ghost) [inline test/unit/pre-start-check.test.ts:104]: **🧹 Nitpick** · Consistent Test Naming

The test names should follow a consistent naming convention for better readability and maintainability.

```suggestion
it('should return go for a clean, unclaimed, actionable issue', () => { ... }
```

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In test/unit/pre-start-check.test.ts around line 104, The test names should follow a consistent naming convention for better readability and maintainability. Apply: it('should return go for a clean, unclaimed, actionable issue', () => { ... }
```

</details>

File: src/mcp/server.ts [hunk 5]
```diff
@@ -69,6 +69,7 @@ import {
   buildLinkedIssueValidation,
   buildLocalDiffPreflightResult,
```
Comment (ghost) [inline src/mcp/server.ts:70]: **🧹 Nitpick** · Duplicate schema definition

The `checkBeforeStartShape` is defined both in this file and in the CLI script, risking divergence.

```suggestion
Extract the schema to a shared module and import it in both places.
```

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/mcp/server.ts around line 70, The `checkBeforeStartShape` is defined both in this file and in the CLI script, risking divergence. Apply: Extract the schema to a shared module and import it in both places.
```

</details>

File: src/signals/engine.ts [hunk 6]
```diff
@@ -2919,6 +2919,190 @@ export function buildLinkedIssueValidation(
...
 
+export type PreStartCheckTarget = {
+  issueNumber?: number | undefined;
+  title?: string | undefined;
+  plannedPaths?: string[] | undefined;
+};
+
+export type PreStartCheckClaimStatus = "unclaimed" | "claimed" | "solved" | "unknown";
+export type PreStartCheckRecommendation = "go" | "raise" | "avoid";
+export type DuplicateClusterRisk = "none" | "low" | "medium" | "high";
```
Comment (ghost) [inline src/signals/engine.ts:2930]: **⚠️ Potential issue** · Potential Edge Case in Title Matching

The title matching logic may not handle very short or very long titles correctly. Ensure that edge cases are tested.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 2930, The title matching logic may not handle very short or very long titles correctly. Ensure that edge cases are tested.
```

</details>

File: src/signals/engine.ts [hunk 7]
```diff
@@ -2919,6 +2919,190 @@ export function buildLinkedIssueValidation(
...
+  const issueClusters =
+    resolvedNumber == null ? [] : collisions.clusters.filter((cluster) => cluster.items.some((item) => item.type === "issue" && item.number === resolvedNumber));
+  // Open PR records carry no file metadata in the cache, so planned-path overlap is evaluated against recently merged work.
+  const pathOverlapMergedPullRequests =
+    plannedPaths.length === 0 ? [] : recentMergedPullRequests.filter((pr) => pr.changedFiles.some((file) => plannedPaths.includes(file.toLowerCase())));
+
+  let duplicateClusterRisk: DuplicateClusterRisk = "none";
+  const riskCandidates: DuplicateClusterRisk[] = [...issueClusters.map((cluster) => cluster.risk), ...(pathOverlapMergedPullRequests.length > 0 ? (["medium"] as const) : [])];
+  for (const risk of riskCandidates) {
+    if (DUPLICATE_RISK_RANK[risk] > DUPLICATE_RISK_RANK[duplicateClusterRisk]) duplicateClusterRisk = risk;
```
Comment (ghost) [inline src/signals/engine.ts:3020]: **⚠️ Potential issue** · Planned Paths Overlap Logic

Verify that the planned paths logic correctly identifies overlaps with recently merged work.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 3020, Verify that the planned paths logic correctly identifies overlaps with recently merged work.
```

</details>

File: src/mcp/server.ts [hunk 8]
```diff
@@ -594,6 +615,17 @@ export class GittensoryMcp {
       async (input) => this.toolResult(await this.validateLinkedIssue(input)),
     );
```
Comment (ghost) [inline src/mcp/server.ts:617]: **⚠️ Potential issue** · Access Control Verification

Double-check the access control logic to ensure that only authorized users can perform pre-start checks.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/mcp/server.ts around line 617, Double-check the access control logic to ensure that only authorized users can perform pre-start checks.
```

</details>

File: src/signals/engine.ts [hunk 9]
```diff
@@ -2919,6 +2919,190 @@ export function buildLinkedIssueValidation(
...
+  };
+  found: boolean;
+  claimStatus: PreStartCheckClaimStatus;
+  lifecycle?: IssueDiscoveryLifecycleState | undefined;
+  issueQualityStatus?: "ready" | "needs_proof" | "hold" | "do_not_use" | undefined;
+  duplicateClusterRisk: DuplicateClusterRisk;
+  recommendation: PreStartCheckRecommendation;
+  reasons: string[];
+  blockers: string[];
+  summary: string;
```
Comment (ghost) [inline src/signals/engine.ts:2950]: **🛠️ Refactor suggestion** · Simplify Title Matching Logic

Consider refactoring the title matching logic to improve readability and maintainability.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 2950, Consider refactoring the title matching logic to improve readability and maintainability.
```

</details>

File: src/signals/engine.ts [hunk 10]
```diff
@@ -2919,6 +2919,190 @@ export function buildLinkedIssueValidation(
...
+      target.issueNumber != null
+        ? `Issue #${target.issueNumber} was not found in cached open-issue metadata; confirm it exists and is open before starting.`
+        : "No matching open issue or overlapping work was found in cached metadata; confirm the target before starting.",
+    );
+  }
+  if (claimStatus === "solved") blockers.push("This issue already has merged or validated solving work; new work would likely duplicate it.");
+  if (claimStatus === "claimed") {
+    blockers.push(
+      resolvedNumber != null
+        ? "Open PR work already references this issue; coordinate or pick a different target to avoid a collision."
```
Comment (ghost) [inline src/signals/engine.ts:3050]: **🛠️ Refactor suggestion** · Optimize Duplicate Cluster Risk Calculation

The duplicate cluster risk calculation can be optimized for better performance.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In src/signals/engine.ts around line 3050, The duplicate cluster risk calculation can be optimized for better performance.
```

</details>

File: test/unit/pre-start-check.test.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+  it("avoids an issue already solved by merged work", () => {
+    const r = repo("owner/repo");
+    const issues = [issue(2, "Improve retry backoff in the queue worker")];
+    const merged = [mergedPr(20, "Improve retry backoff", { linkedIssues: [2] })];
+    const report = buildPreStartCheck(r, issues, [], merged, "owner/repo", { issueNumber: 2 });
+    expect(report.claimStatus).toBe("solved");
+    expect(report.lifecycle).toBe("valid_solved");
+    expect(report.recommendation).toBe("avoid");
+    expect(report.blockers.some((b) => /merged or validated/i.test(b))).toBe(true);
```
Comment (ghost) [inline test/unit/pre-start-check.test.ts:120]: **🧹 Nitpick** · Consistent Test Naming

Ensure consistent naming conventions for test cases to improve readability.

<details>
<summary>🤖 Prompt for AI agents</summary>

```
In test/unit/pre-start-check.test.ts around line 120, Ensure consistent naming conventions for test cases to improve readability.
```

</details>

**Review by ghost — COMMENTED** [review]
**Gittensory review** · _advisory — a maintainer merges_

**5** actionable · **7** nitpick(s) · 8 file(s) · two independent AI reviewers

> **Suggested action:** ✅ **Safe to merge** — both reviewers found no blocking issues.

<details>
<summary>📋 Walkthrough</summary>

The PR adds a new `gittensory_check_before_start` tool that evaluates issue claim status, duplicate risk, and planned path overlap before any code is written, exposing a safe public recommendation via a new API route and MCP registration, with full test coverage.

**Changes**

| File | Summary |
| --- | --- |
| `packages/gittensory-mcp/bin/gittensory-mcp.js` | Defines input schema for the new pre‑start check tool and registers it with the CLI. |
| `src/api/routes.ts` | Adds `/check-before-start` POST endpoint, validates input, authorizes, and returns the pre‑start check report. |
| `src/mcp/server.ts` | Registers the new tool in MCP server, adds input/output schemas, and implements the check logic. |
| `src/signals/engine.ts` | Implements `buildPreStartCheck` function with detailed risk analysis and public‑safe output. |
| `test/integration/api.test.ts` | Tests successful, invalid, and forbidden scenarios for the new API endpoint. |
| `test/unit/mcp-output-schemas.test.ts` | Ensures the tool's output conforms to the defined schema. |
| `test/unit/mcp-upstream.test.ts` | Verifies session access restrictions for the pre‑start check. |
| `test/unit/pre-start-check.test.ts` | Adds extensive unit tests for `buildPreStartCheck` covering various edge cases and public safety. |

</details>

<details>
<summary>🔍 Reviewer notes</summary>

**Reviewer A · `gpt-oss-120b`** — recommends ✅ `merge`
This PR introduces a new pre‑start duplicate/solvability check tool, adding API endpoints, MCP registration, implementation, type definitions, and comprehensive tests. The changes are well‑scoped, safe, and respect the public‑private sanitizer.

_Suggestions_
- Consider consolidating the duplicated `checkBeforeStartShape` definition (present in both the CLI and MCP server) to a shared schema to avoid future inconsistencies.
- Add a unit test for the case where both `title` and `plannedPaths` are supplied to ensure the matching logic prioritises correctly.

_Worth double-checking_
- Ensure that any future modifications keep the public‑safe sanitization on all string fields returned by the tool.
- Verify that the input schema limits (e.g., title length) stay consistent between the CLI and server definitions.

**Reviewer B · `mistral-small-3.1-24b-instruct`** — recommends ✅ `merge`
This PR adds a new MCP tool and corresponding API route for pre-start checks on GitHub issues. It includes comprehensive validation, detailed reporting, and thorough testing. The implementation is clean and adheres to the project's conventions.

_Suggestions_
- Consider adding a more detailed description in the PR title to clarify the purpose and impact of the new feature.
- Ensure that the new tool and API route are documented in the project's documentation.
- Review the test coverage to ensure all edge cases are handled, especially around title matching and planned paths.

_Worth double-checking_
- Verify that the title matching logic handles edge cases, such as very short or very long titles.
- Ensure that the planned paths logic correctly identifies overlaps with recently merged work.
- Double-check the access control logic to ensure that only authorized users can perform pre-start checks.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

Copy this to your coding agent to fix everything flagged:

```
Address the following in this PR:

1. src/mcp/server.ts (line 70): The `checkBeforeStartShape` is defined both in this file and in the CLI script, risking divergence. Suggested fix: Extract the schema to a shared module and import it in both places.
2. src/signals/engine.ts (line 2930): The title matching logic may not handle very short or very long titles correctly. Ensure that edge cases are tested.
3. src/signals/engine.ts (line 3020): Verify that the planned paths logic correctly identifies overlaps with recently merged work.
4. src/mcp/server.ts (line 617): Double-check the access control logic to ensure that only authorized users can perform pre-start checks.
5. src/signals/engine.ts (line 2950): Consider refactoring the title matching logic to improve readability and maintainability.
6. src/signals/engine.ts (line 3050): The duplicate cluster risk calculation can be optimized for better performance.
7. test/unit/pre-start-check.test.ts (line 120): Ensure consistent naming conventions for test cases to improve readability.

Nitpicks:
- Consider consolidating the duplicated `checkBeforeStartShape` definition (present in both the CLI and MCP server) to a shared schema to avoid future inconsistencies.
- Add a unit test for the case where both `title` and `plannedPaths` are supplied to ensure the matching logic prioritises correctly.
- Consider adding a more detailed description in the PR title to clarify the purpose and impact of the new feature.
- Ensure that the new tool and API route are documented in the project's documentation.
- Review the test coverage to ensure all edge cases are handled, especially around title matching and planned paths.
```

</details>

---
_🤖 Automated advisory review · a maintainer makes the final call._

**Review by JSONbored — APPROVED** [review]
